File: python/aethermesh/content/chunk_bitmap.py
# SPDX-License-Identifier: MIT
"""
ChunkBitmap wire-format codec for the Aether Chunk Shuffle / SAPI protocol.

Wire format:
  • JSON, snake_case property names.
  • Bitset: LSB-first within each byte — bit i is set in byte (i//8), at
    position (i%8).  Length = ceil(chunk_count / 8).
  • Bitset transmitted as standard Base64 (with padding).
  • Field order in canonical JSON: root_hash, chunk_count, have_bitset,
    generation.
"""

import base64
import json
import math
from typing import Iterable


class BitsetCodec:
    @staticmethod
    def encode(chunk_count: int, have_indices: Iterable[int]) -> bytes:
        """Encode indices into an LSB-first compact bitset.

        Returns bytes of length ceil(chunk_count / 8). Raises ValueError if
        any index is out of [0, chunk_count).
        """
        if chunk_count <= 0:
            return b""
        buf = bytearray((chunk_count + 7) // 8)
        for i in have_indices:
            if i < 0 or i >= chunk_count:
                raise ValueError(f"Index {i} out of range [0, {chunk_count})")
            buf[i >> 3] |= (1 << (i & 7))
        return bytes(buf)

    @staticmethod
    def decode(bitset: bytes, chunk_count: int) -> list[int]:
        """Decode a compact bitset into sorted chunk indices."""
        result = []
        limit = min(chunk_count, len(bitset) * 8)
        for i in range(limit):
            if bitset[i >> 3] & (1 << (i & 7)):
                result.append(i)
        return result


def marshal_json_chunk_bitmap(
    root_hash: str,
    chunk_count: int,
    have_bitset: bytes,
    generation: int,
) -> str:
    """Produce canonical wire JSON with fixed field order."""
    have_b64 = base64.b64encode(have_bitset).decode("ascii")
    # Use json.dumps for values but assemble field order manually
    parts = [
        f'"root_hash":{json.dumps(root_hash)}',
        f'"chunk_count":{chunk_count}',
        f'"have_bitset":{json.dumps(have_b64)}',
        f'"generation":{generation}',
    ]
    return "{" + ",".join(parts) + "}"

File: swift/Sources/AetherNetProtocol/Content/ChunkBitmap.swift
// SPDX-License-Identifier: MIT
//
// ChunkBitmapPayload wire-format codec — Swift implementation.
//
// Compact LSB-first bitset encoding for the ChunkBitmap broadcast packet
// (PacketType 37). Bit i is set in byte (i>>3) at bit-position (i&7).
// Output length is exactly ceil(chunk_count / 8) bytes; trailing bits are
// always zero.
//
// Cross-language stable: the same bit-packing and Base64-with-padding
// encoding is implemented in C, C#, Go, Python, TypeScript, Rust, Kotlin.
// The canonical JSON field order is:
//   root_hash, chunk_count, have_bitset, generation

import Foundation

// ── BitsetCodec ────────────────────────────────────────────────────────────────

/// Compact LSB-first bitset codec for ChunkBitmapPayload.
public enum BitsetCodec {

    /// Encode a list of present-chunk indices into an LSB-first compact bitset.
    ///
    /// - Parameters:
    ///   - chunkCount: Total number of chunks in the content (≥ 0).
    ///   - haveIndices: Chunk indices that are present. Out-of-range values
    ///     are silently ignored.
    /// - Returns: `Data` of length `ceil(chunkCount / 8)`, or empty when
    ///   `chunkCount == 0`.
    public static func encode(chunkCount: Int, haveIndices: [Int]) -> Data {
        guard chunkCount > 0 else { return Data() }
        let byteLen = (chunkCount + 7) / 8
        var bytes = Data(repeating: 0, count: byteLen)
        for idx in haveIndices {
            guard idx >= 0 && idx < chunkCount else { continue }
            bytes[idx >> 3] |= UInt8(1 << (idx & 7))
        }
        return bytes
    }

    /// Decode a compact bitset into a sorted array of set-bit indices.
    ///
    /// - Parameters:
    ///   - bitset: Compact bitset bytes.
    ///   - chunkCount: Total chunk count; bits beyond this limit are ignored.
    /// - Returns: Sorted `[Int]` of set-bit indices. Empty when no bits are
    ///   set or `bitset` is empty.
    public static func decode(bitset: Data, chunkCount: Int) -> [Int] {
        guard !bitset.isEmpty && chunkCount > 0 else { return [] }
        let limit = min(chunkCount, bitset.count * 8)
        var result: [Int] = []
        for i in 0..<limit where (bitset[i >> 3] & UInt8(1 << (i & 7))) != 0 {
            result.append(i)
        }
        return result
    }
}

// ── JSON marshal ───────────────────────────────────────────────────────────────

/// Produce the canonical wire JSON for a ChunkBitmapPayload:
///
///     {"root_hash":"...","chunk_count":N,"have_bitset":"<base64>","generation":G}
///
/// Field order is fixed by the cross-language specification.
/// Base64 encoding uses RFC 4648 §4 (standard alphabet, padded) via
/// `Data.base64EncodedString()`.
///
/// - Parameters:
///   - rootHash: Lowercase hex SHA-256 root hash (no JSON-special characters).
///   - chunkCount: Total chunk count.
///   - haveBitset: Bitset bytes returned by `BitsetCodec.encode`.
///   - generation: Generation counter (`UInt32`).
/// - Returns: Canonical JSON string.
public func marshalChunkBitmapJson(
    rootHash: String,
    chunkCount: Int,
    haveBitset: Data,
    generation: UInt32
) -> String {
    let b64 = haveBitset.base64EncodedString()
    // Single-line concatenation guarantees no embedded newlines in the output.
    return "{\"root_hash\":\"\(rootHash)\",\"chunk_count\":\(chunkCount),"
         + "\"have_bitset\":\"\(b64)\",\"generation\":\(generation)}"
}
